jwst/master_background/expand_to_2d.py

```diff
@@ -0,0 +1,314 @@
+
+    tab_background : 1-D ndarray
+        The flux column read from the 1-D background table, divided by
+        the npixels column.
+
+    Returns
+    -------
+    background : JWST IFUImageModel
+        A copy of `input` but with the data replaced by the background,
+        "expanded" from 1-D to 2-D.
+    """
+
+    background = input.copy()
+    background.data[:, :] = 0.
+
+    if input.meta.instrument.name == "NIRSPEC":
+        list_of_wcs = nirspec.nrs_ifu_wcs(input)
+        for ifu_wcs in list_of_wcs:
+
+            xstart = ifu_wcs.bounding_box[0][0]
+            xstop = ifu_wcs.bounding_box[0][1]
+            ystart = ifu_wcs.bounding_box[1][0]
+            ystop = ifu_wcs.bounding_box[1][1]
+
+            # Convert to integers, and add one to the upper limits,
+            # because we want to use these as slice limits.
+            xstart = math.floor(xstart)
+            xstop = math.ceil(xstop) + 1
+            ystart = math.floor(ystart)
+            ystop = math.ceil(ystop) + 1
+            xstart = max(xstart, 0)
+            ystart = max(ystart, 0)
+            xstop = min(xstop, 2048)
+            ystop = min(ystop, 2048)
+
+            shape = (ystop - ystart, xstop - xstart)
+            grid = np.indices(shape, dtype=np.float64)
+            grid[0] += ystart
+            grid[1] += xstart
+            wl_array = ifu_wcs(grid[1], grid[0])[2]
+
+            wl_array[np.isnan(wl_array)] = -1.
+            bkg_flux = np.interp(wl_array, tab_wavelength, tab_background,
+                                 left=0., right=0.)
+            background.data[ystop:ystart, xstop:xstart] = bkg_flux.copy()
+
+    elif input.meta.instrument.name == "MIRI":
+        shape = input.data.shape
+        grid = np.indices(shape, dtype=np.float64)
+        wl_array = ifu_wcs(grid[1], grid[0])[2]
+
+        wl_array[np.isnan(wl_array)] = -1.
+        bkg_flux = np.interp(wl_array, tab_wavelength, tab_background,
+                             left=0., right=0.)
+        background.data[:, :] = bkg_flux.copy()
+
+    else:
+        raise RuntimeError("Exposure type {} is not supported."
+                           .format(input.meta.exposure.type))
+
+    return background
+
+
+def get_wavelengths(model):
+    """Read or compute wavelengths.
+
+    Parameters
+    ----------
+    model : JWST data model
+        The input science data.
+
+    Returns
+    -------
+    wl_array : 2-D ndarray
+        An array of wavelengths corresponding to the data in `model`.
+    """
+
+    if hasattr(model, "wavelength"):
+        wl_array = model.wavelength.copy()
+        got_wavelength = True                   # may be reset below
+    else:
+        wl_array = None
+    if (wl_array is None or len(wl_array) == 0 or
+        np.nanmin(wl_array) == 0. and np.nanmax(wl_array) == 0.):
+            got_wavelength = False
+            wl_array = None
+
+    if hasattr(model, "wcs") and not got_wavelength:
```

Comment:
"wcs" is an attribute of model.meta. `model.wcs` points to the FITS WCS which we don't want to use.
  This needs to be fixed but is a separate issue.

Reply:
Thanks!  I wonder whether I made this mistake in other steps, too.